validator.go

```diff
@@ -132,6 +137,87 @@ func runValidators(val interface{}, validators []validatorTag) error {
 	return nil
 }
 
+func tryValidateValue(val reflect.Value, opts *options, validators []validatorTag) error {
+	var curr interface{}
+	if val.IsValid() {
+		curr = val.Interface()
+	}
+	if err := runValidators(curr, validators); err != nil {
+		return err
+	}
+	if !val.IsValid() {
+		return nil
+	}
+
+	var err error
+	switch chaseValue(val).Kind() {
+	case reflect.Struct:
+		err = validateStruct(val, opts)
+	case reflect.Map:
+		err = validateMap(val, opts)
+	case reflect.Array, reflect.Slice:
+		err = validateArray(val, opts)
+	}
+
+	if err != nil {
+		return err
+	}
+	return tryValidate(val)
+}
+
+func validateStruct(val reflect.Value, opts *options) error {
+	val = chaseValue(val)
+	numField := val.NumField()
+	for i := 0; i < numField; i++ {
+		stField := val.Type().Field(i)
+
+		// ignore non exported fields
+		if rune, _ := utf8.DecodeRuneInString(stField.Name); !unicode.IsUpper(rune) {
+			continue
+		}
+		_, tagOpts := parseTags(stField.Tag.Get(opts.tag))
+		if tagOpts.ignore {
+			continue
+		}
```

Comment:
Looks like we have this block of code in at least two places now. Let's unify in case we want ehnahce/fix something on how to access/skip fields or update options:

e.g.
```
func accessField(structVal reflect.Value, fieldIdx int) (field reflect.Value, *options, ignore bool) {
   ...
}
```